### PR TITLE
`Tutorial groups`: Give students information about the current utilization and average attendance

### PR DESCRIPTION
**Pull request by stefanwaldhauser**
#### General
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [X] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [X] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [X] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [X] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [X] I implemented the changes with a good performance and prevented too many database calls.
- [X] I documented the Java code using JavaDoc style.
#### Client
- [X] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.
- [X] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [X] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [X] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [X] I documented the TypeScript code using JSDoc style.
- [X] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context
- We want to show the average attendance and utilization of a tutorial group, so students know which are crowded groups and which are empty groups where a more personal tutoring is possible.

### Description
- Added the `averageAttendance` transient property to the `tutorialGroup` entiy. It is calculated in the following way:
```
   /**
     * Sets the averageAttendance transient field of the given tutorial group
     * <p>
     * Calculation:
     * <ul>
     *     <li>Get set of the last three completed sessions (or less than three if not more available) </li>
     *     <li>Remove sessions without attendance data (null) from the set</li>
     *     <li>If set is empty, set attendance average of tutorial group to null (meaning could not be determined)</li>
     *     <li>If set is non empty, set the attendance average of the tutorial group to the arithmetic mean</li>
     * </ul>
     *
     * @param tutorialGroup the tutorial group to set the averageAttendance for
     */
```

Now on the client I have added two new columns to the tutorial group table. One for the `averageAttendance` and one for the `utilization` which is calculated by `averageAttendance/capacity`. The `utilization` is displayed via a vertical progress bar. Help texts explain the calcultion via tooltips.

### Steps for Testing
- Create a tutorial group with at least three sessions in the past. (with capacity set)
- Add attendance data to these sessions.
- Check that the average attendance and utilization is correctly calculated.

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage
- As you can see from the code I have added both server and client tests to cover the changes.

| Component                                         | Statement Coverage | Branch Coverage |
|---------------------------------------------------|--------------------|-----------------|
| vertical-progress-bar.component.ts                | 97.56%             | 90.47%          |
| tutorial-group-utilization-indicator.component.ts | 100%               | 100%            |
| TutorialGroupService.java                         | 94%                | 74%             |

### Screenshots
image
image
image

### CODE REVIEW

File: src/main/webapp/app/course/tutorial-groups/shared/tutorial-group-detail/tutorial-group-detail.component.ts [hunk 1]
```diff
@@ -32,6 +33,9 @@ export class TutorialGroupDetailComponent implements OnChanges {
     formattedAdditionalInformation?: SafeHtml;
     getDayTranslationKey = getDayTranslationKey;
 
+    faQuestionCircle = faQuestionCircle;
+    math = Math;
```
Comment (Strohgelaender) [inline src/main/webapp/app/course/tutorial-groups/shared/tutorial-group-detail/tutorial-group-detail.component.ts:37]: same for the other classes
```suggestion
    readonly Math = Math;
```

**Review by Strohgelaender — COMMENTED** [review]
Is it just me or does the attendence meeter look like a beer? 🍺

~~I'd sugesst to use different colors (green, yellow, red) depending on how high the value is.~~ Looks like that's already implemented. Can you add additional screenshots to the PR description?

What about overcroweded groups where students have to stand? Is this possible / is there a represenation for this?

**Comment by stefanwaldhauser** [comment]
> Is it just me or does the attendence meeter look like a beer? 🍺
> 
> ~I'd sugesst to use different colors (green, yellow, red) depending on how high the value is.~ Looks like that's already implemented. Can you add additional screenshots to the PR description?
> 
> What about overcroweded groups where students have to stand? Is this possible / is there a represenation for this?

I think the beer thing is on you :P. There is currently no representation for the standing issue you described. We want to keep it simple in this PR

File: src/main/webapp/app/course/tutorial-groups/shared/tutorial-group-detail/tutorial-group-detail.component.ts (hunk 1 above)
Comment (stefanwaldhauser) [reply src/main/webapp/app/course/tutorial-groups/shared/tutorial-group-detail/tutorial-group-detail.component.ts]: Done in 5a44a6571d5bb5b56e26be52ddc3fb6e25c3276e

File: src/main/webapp/app/shared/vertical-progress-bar/vertical-progress-bar.component.ts [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+ *     <li> [0,lowerBorder] -> lower color
+ * </ul>
+ */
+@Component({
+    selector: 'jhi-vertical-progress-bar',
+    templateUrl: './vertical-progress-bar.component.html',
+    styleUrls: ['./vertical-progress-bar.component.scss'],
+    changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class VerticalProgressBarComponent implements OnInit {
```
Comment (ge65cer) [inline src/main/webapp/app/shared/vertical-progress-bar/vertical-progress-bar.component.ts:18]: Can you elaborate why you chose a custom vertical implementation over the default Bootstrap progress bars?
https://ng-bootstrap.github.io/#/components/progressbar/examples

File: src/main/webapp/app/course/tutorial-groups/shared/tutorial-group-detail/tutorial-group-detail.component.html [hunk 3]
```diff
@@ -38,6 +38,37 @@ <h3>{{ 'artemisApp.pages.tutorialGroupDetail.title' | artemisTranslate }}:</h3>
...
         </dd>
+        <!-- Utilization -->
+        <dt>
+            <span>{{ 'artemisApp.entities.tutorialGroup.utilization' | artemisTranslate }}</span>
+            <fa-icon [icon]="faQuestionCircle" class="text-secondary" ngbTooltip="{{ 'artemisApp.entities.tutorialGroup.utilizationHelp' | artemisTranslate }}"></fa-icon>
+        </dt>
+        <dd>
+            <ng-container *ngIf="tutorialGroup.averageAttendance !== undefined && tutorialGroup.capacity !== undefined; else noUtilization">
+                <jhi-vertical-progress-bar
+                    [fillLevelInPercent]="Math.round((tutorialGroup.averageAttendance / tutorialGroup.capacity) * 100)"
```
Comment (ge65cer) [inline src/main/webapp/app/course/tutorial-groups/shared/tutorial-group-detail/tutorial-group-detail.component.html:49]: Might be better to calculate this inside the component (computed property) and not the template.

File: src/main/webapp/app/course/tutorial-groups/shared/tutorial-group-detail/tutorial-group-detail.component.html [hunk 4]
```diff
@@ -38,6 +38,37 @@ <h3>{{ 'artemisApp.pages.tutorialGroupDetail.title' | artemisTranslate }}:</h3>
...
+            </ng-template>
+        </dd>
+        <!-- Average Attendance -->
+        <dt>
+            <span>{{ 'artemisApp.entities.tutorialGroup.averageAttendance' | artemisTranslate }}</span>
+            <fa-icon [icon]="faQuestionCircle" class="text-secondary" ngbTooltip="{{ 'artemisApp.entities.tutorialGroup.averageAttendanceHelp' | artemisTranslate }}"></fa-icon>
+        </dt>
+        <dd>
+            <ng-container *ngIf="tutorialGroup.averageAttendance !== undefined; else noAverageAttendance">
+                <span>{{ Math.round(tutorialGroup.averageAttendance) }}</span>
```
Comment (ge65cer) [inline src/main/webapp/app/course/tutorial-groups/shared/tutorial-group-detail/tutorial-group-detail.component.html:66]: Same here, so we don't need to create `Math` reference.

File: src/main/webapp/app/shared/vertical-progress-bar/vertical-progress-bar.component.ts (hunk 2 above)
Comment (ge65cer) [reply src/main/webapp/app/shared/vertical-progress-bar/vertical-progress-bar.component.ts:18]: To add to that, the Rating component (e.g., with a student icon) could also work great here. 
https://ng-bootstrap.github.io/#/components/rating/examples#decimal

**Review by timo07k — APPROVED** [review]
Works as expected. 
Small remark: Should it be possible to type in higher attendance than capacity?

image

Tested on TS1.

File: src/main/webapp/app/shared/vertical-progress-bar/vertical-progress-bar.component.ts (hunk 2 above)
Comment (stefanwaldhauser) [reply src/main/webapp/app/shared/vertical-progress-bar/vertical-progress-bar.component.ts:18]: I just wanted to play around a little bit with css. See no harm in it

File: src/main/java/de/tum/in/www1/artemis/service/tutorialgroups/TutorialGroupService.java [hunk 5]
```diff
@@ -97,6 +97,40 @@ public void setTransientPropertiesForUser(User user, TutorialGroup tutorialGroup
...
+     * <li>Remove sessions without attendance data (null) from the set</li>
+     * <li>If set is empty, set attendance average of tutorial group to null (meaning could not be determined)</li>
+     * <li>If set is non empty, set the attendance average of the tutorial group to the arithmetic mean (rounded to integer)</li>
+     * </ul>
+     *
+     * @param tutorialGroup the tutorial group to set the averageAttendance for
+     */
+    private void setAverageAttendance(TutorialGroup tutorialGroup) {
+        Collection<TutorialGroupSession> sessions;
+        if (getPersistenceUtil().isLoaded(tutorialGroup, "tutorialGroupSessions") && tutorialGroup.getTutorialGroupSessions() != null) {
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/tutorialgroups/TutorialGroupService.java:118]: Why don't you use Hibernate.isInitialized(), that way future refactoring does not have to care about these string literals

Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/service/tutorialgroups/TutorialGroupService.java:118]: Persistence.getPersistenceUtil is just more generic as if you can change JPA implementation underneath without changing your code.

File: src/main/webapp/app/course/tutorial-groups/shared/tutorial-group-detail/tutorial-group-detail.component.html (hunk 3 above)
Comment (stefanwaldhauser) [reply src/main/webapp/app/course/tutorial-groups/shared/tutorial-group-detail/tutorial-group-detail.component.html:49]: Personally I think this is really a matter of taste tbh.

File: src/main/webapp/app/course/tutorial-groups/shared/tutorial-group-detail/tutorial-group-detail.component.html (hunk 4 above)
Comment (stefanwaldhauser) [reply src/main/webapp/app/course/tutorial-groups/shared/tutorial-group-detail/tutorial-group-detail.component.html:66]: Personally I think this is really a matter of taste tbh.

**Review by LiaLism — APPROVED** [review]
Utilization and average attendance worked as expected. Tested on ts0.

**Review by PhilipWerz — DISMISSED** [review]
Everything works as expected. Attendence and utilization are calculated correctly. Tested on TS1.

**Review by luisa1999 — APPROVED** [review]
Looks good to me too!

**Review by grochlitzzz — APPROVED** [review]
Everything working as expected. Tested with less than 3 sessions as well and average keeps getting updated. Also correctly displayed for tutorial group participants. Tested on ts2

File: src/main/java/de/tum/in/www1/artemis/domain/tutorialgroups/TutorialGroup.java [hunk 6]
```diff
@@ -97,6 +97,14 @@ public class TutorialGroup extends DomainObject {
     @Transient
     private TutorialGroupSession nextSession;
 
+    /**
+     * This field represents the average attendance of this tutorial group
+     * <p>
+     * For more information on how this is calculated check out {@link de.tum.in.www1.artemis.service.tutorialgroups.TutorialGroupService#setAverageAttendance}
+     */
+    @Transient
+    private Integer averageAttendance;
```
Comment (simon-hng) [inline src/main/java/de/tum/in/www1/artemis/domain/tutorialgroups/TutorialGroup.java:106]: Is there a more fitting name? I am not sure what to expect with an `Integer` called `averageAttendance`. I am sure there is a reason behind this but I would expect a float/double when reading "average"

File: src/main/webapp/app/shared/vertical-progress-bar/vertical-progress-bar.component.html [hunk 7]
```diff
@@ -0,0 +1,6 @@
+<div class="vertical-progress-bar">
```
Comment (simon-hng) [inline src/main/webapp/app/shared/vertical-progress-bar/vertical-progress-bar.component.html:1]: why don't we use ngxcharts here?

**Review by simon-hng — DISMISSED** [review]
Left minor suggestions. Code looks overall good

File: src/main/webapp/app/shared/vertical-progress-bar/vertical-progress-bar.component.html (hunk 7 above)
Comment (stefanwaldhauser) [reply src/main/webapp/app/shared/vertical-progress-bar/vertical-progress-bar.component.html:1]: Huge performance overhead for a simple progress bar

File: src/main/java/de/tum/in/www1/artemis/service/tutorialgroups/TutorialGroupService.java [hunk 8]
```diff
@@ -97,6 +97,40 @@ public void setTransientPropertiesForUser(User user, TutorialGroup tutorialGroup
...
+            sessions = tutorialGroup.getTutorialGroupSessions();
+        }
+        else {
+            sessions = tutorialGroupSessionRepository.findAllByTutorialGroupId(tutorialGroup.getId());
+        }
+        sessions.stream()
+                .filter(tutorialGroupSession -> TutorialGroupSessionStatus.ACTIVE.equals(tutorialGroupSession.getStatus())
+                        && tutorialGroupSession.getEnd().isBefore(ZonedDateTime.now()))
+                .sorted(Comparator.comparing(TutorialGroupSession::getStart).reversed()).limit(3)
```
Comment (bassner) [inline src/main/java/de/tum/in/www1/artemis/service/tutorialgroups/TutorialGroupService.java:127]: (optional) I think the limit step can easily be overlooked as all the other steps follow a vertical flow
```suggestion
                .sorted(Comparator.comparing(TutorialGroupSession::getStart).reversed())
                .limit(3)
```

File: src/main/webapp/app/shared/vertical-progress-bar/vertical-progress-bar.component.ts [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+        this.upperBorderInternal = Math.round(Math.min(border, 100));
+        this.setFillColor();
+        this.cdr.markForCheck;
+    }
+    @Input()
+    lowerColor = 'green';
+    @Input()
+    intermediateColor = 'orange';
+    @Input()
+    upperColor = 'red';
```
Comment (bassner) [inline src/main/webapp/app/shared/vertical-progress-bar/vertical-progress-bar.component.ts:62]: These colors are not theme-aware, hard-coded colors. Please use default colors from the theme, in this case that would be ``success``, ``warning`` and ``danger``. CSS variables for these colors are globally available

File: src/test/javascript/spec/component/shared/vertical-progress-bar.component.spec.ts [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+        }).compileComponents();
+
+        fixture = TestBed.createComponent(VerticalProgressBarComponent);
+        component = fixture.componentInstance;
+        fixture.detectChanges();
+    });
+
+    it('should create', () => {
+        expect(component).toBeTruthy();
+    });
```
Comment (bassner) [inline src/test/javascript/spec/component/shared/vertical-progress-bar.component.spec.ts:22]: Does this test serve any purpose? The other test will fail, too, if the component can't be created. :)

**Review by bassner — CHANGES_REQUESTED** [review]
Hey, I've added my code review below!

File: src/main/webapp/app/course/tutorial-groups/shared/tutorial-group-detail/tutorial-group-detail.component.html [hunk 11]
```diff
@@ -38,6 +38,37 @@ <h3>{{ 'artemisApp.pages.tutorialGroupDetail.title' | artemisTranslate }}:</h3>
...
+                <jhi-vertical-progress-bar
+                    [fillLevelInPercent]="Math.round((tutorialGroup.averageAttendance / tutorialGroup.capacity) * 100)"
+                    [widthInPixels]="20"
+                    [heightInPixels]="50"
+                >
+                </jhi-vertical-progress-bar>
+            </ng-container>
+            <ng-template #noUtilization>
+                <span>{{ '' }}</span>
+            </ng-template>
```
[the remark this answers is not in the document]
Comment (stefanwaldhauser) [reply src/main/webapp/app/course/tutorial-groups/shared/tutorial-group-detail/tutorial-group-detail.component.html]: Done in 4a7932b6a7b50b48226a6c531a76fb9c76fa4b88

File: src/main/webapp/app/course/tutorial-groups/shared/tutorial-group-detail/tutorial-group-detail.component.html [hunk 12]
```diff
@@ -38,6 +38,37 @@ <h3>{{ 'artemisApp.pages.tutorialGroupDetail.title' | artemisTranslate }}:</h3>
...
+            <span>{{ 'artemisApp.entities.tutorialGroup.averageAttendance' | artemisTranslate }}</span>
+            <fa-icon [icon]="faQuestionCircle" class="text-secondary" ngbTooltip="{{ 'artemisApp.entities.tutorialGroup.averageAttendanceHelp' | artemisTranslate }}"></fa-icon>
+        </dt>
+        <dd>
+            <ng-container *ngIf="tutorialGroup.averageAttendance !== undefined; else noAverageAttendance">
+                <span>{{ Math.round(tutorialGroup.averageAttendance) }}</span>
+            </ng-container>
+            <ng-template #noAverageAttendance>
+                <span>{{ '' }}</span>
+            </ng-template>
```
[the remark this answers is not in the document]
Comment (stefanwaldhauser) [reply src/main/webapp/app/course/tutorial-groups/shared/tutorial-group-detail/tutorial-group-detail.component.html:60]: Done in 4a7932b6a7b50b48226a6c531a76fb9c76fa4b88

File: src/test/javascript/spec/component/shared/vertical-progress-bar.component.spec.ts (hunk 10 above)
Comment (stefanwaldhauser) [reply src/test/javascript/spec/component/shared/vertical-progress-bar.component.spec.ts]: Removed in 1b0345999922a4ab85d31a1f707a15344fadab9e

File: src/main/webapp/app/shared/vertical-progress-bar/vertical-progress-bar.component.ts (hunk 9 above)
Comment (stefanwaldhauser) [reply src/main/webapp/app/shared/vertical-progress-bar/vertical-progress-bar.component.ts]: Done in 491adfb2516522167fc956c9433b5ebc9e81954e

**Review by grochlitzzz — CHANGES_REQUESTED** [review]
If you edit something in the tutorial configuration (like tutor assigned, additional information, tutorial language), the attendance information from the sessions vanishes. I guess this wasn't supposed to work like this, since some of these fields might change during the semester (like a change in tutor or new information being added to the "additional info" section)

**Review by grochlitzzz — CHANGES_REQUESTED** [review]
> > If you edit something in the tutorial configuration (like tutor assigned, additional information, tutorial language), the attendance information from the sessions vanishes. I guess this wasn't supposed to work like this, since some of these fields might change during the semester (like a change in tutor or new information being added to the "additional info" section)
> 
> fixed here #6236

Apparently still not working

File: src/main/webapp/app/course/tutorial-groups/shared/tutorial-group-detail/tutorial-group-detail.component.html (hunk 3 above)
Comment (ge65cer) [reply src/main/webapp/app/course/tutorial-groups/shared/tutorial-group-detail/tutorial-group-detail.component.html]: What do the other reviewers say? @simon-hng @bassner @Strohgelaender

File: src/main/webapp/app/shared/vertical-progress-bar/vertical-progress-bar.component.ts (hunk 2 above)
Comment (ge65cer) [reply src/main/webapp/app/shared/vertical-progress-bar/vertical-progress-bar.component.ts:18]: What do the other reviewers say? @simon-hng @Strohgelaender

**Review by ge65cer — COMMENTED** [review]
Would be great to get a second opinion on my code review comments above.

**Review by TheZoker — APPROVED** [review]
Tested in testing session and works as expected. Only question we had is, if rounding the average attendance is necessary here?

**Review by sven0311 — APPROVED** [review]
Tested on TS1. Looks good. Only the rounding is a bit misleading

**Review by TimOrtel — APPROVED** [review]
Tested on test server 1, behaviour and calculations look fine to me.

**Review by danielbode — APPROVED** [review]
Tested on ts1. Works as expected. But one remark:

- If a tutorial group is not that big it might be too much to round the average attendance to an integer. I think one decimal would fit better in such a case and wouldn't also be too much in a bigger group

**Review by pellumbbaboci — APPROVED** [review]
Tested during testing session on ts1 works as expected

**Pull request merged**